src/generic/Matrix.jl

```diff
@@ -28,32 +28,23 @@ export MatrixSpace, fflu!, fflu, solve_triu, isrref, charpoly_danilevsky!,
 #
 ###############################################################################
 
-function similar(x::Mat{T}, R::Ring=base_ring(x)) where T <: RingElement
-   TT = elem_type(R)
-   M = similar(x.entries, TT)
-   for i in 1:size(M, 1)
-      for j in 1:size(M, 2)
-         M[i, j] = zero(R)
-      end
-   end
-   z = MatSpaceElem{TT}(M)
-   z.base_ring = R
-   return z
-end
-
-function similar(x::Mat{T}, R::Ring, r::Int, c::Int) where T <: RingElement
+function _similar(x::Union{Mat{T},MatAlgElem{T}}, R::Ring, r::Int, c::Int) where T <: RingElement
    TT = elem_type(R)
    M = similar(x.entries, TT, r, c)
    for i in 1:size(M, 1)
       for j in 1:size(M, 2)
          M[i, j] = zero(R)
       end
    end
-   z = MatSpaceElem{TT}(M)
+   z = x isa Mat ? MatSpaceElem{TT}(M) : MatAlgElem{TT}(M)
    z.base_ring = R
    return z
 end
 
+similar(x::Mat, R::Ring, r::Int, c::Int) = _similar(x, R, r, c)
+
+similar(x::Mat, R::Ring=base_ring(x))= similar(x, R, nrows(x), ncols(x))
```

Comment:
Any reason to not call the _similar method directly?

Reply:
Personal preference, I like to minimize calls to the lower level interface.

Reply:
That's fine, but you are making the compiler work harder, as it has to inline one of the function calls in order to be as fast.

Reply:
I know it works (a bit?) harder, I believed this was worth it when it benefits the programmer. Also as here we allocate a matrix, even if it was not inlined, I guess the overhead of non-inlining would be negligible.

Reply:
Ok

Reply:
@wbhart: this allowed to implement `similar` in Nemo by just defining the general method `similar(mat, ring, r, c)`. If in the method above we had called `_similar`, which is the implementation for generic matrices in `AbstractAlgebra`, it would have worked only for those matrices, and it would have had to be duplicated for other (Nemo, etc.) matrices.

Reply:
Yes, we can probably make use of the same in Singular.jl.